### PR TITLE
5991 - Update ScribeJava to 6.6.3 and necessary refactoring

### PR DESCRIPTION
**Pull request by poikilotherm**
This PR aims to provide a first, smaller step by updating the ScribeJava OAuth2 lib to a recent version.

## Related Issues

- #5974 : Attach dataverse to IDM/IAM
- #5991 : implementing a minimal OpenID connect compliant provider

## Pull Request Checklist

- [x] Unit [tests][] completed
- [x] Integration [tests][]: None
- [x] [Documentation][docs] completed
- [x] Merged latest from "develop" [branch][] and resolved conflicts

## Notable changes for QA (WIP, more to come):
- Removed `scope` attribute from `OAuth2TokenData` in 81b367b7. See commit comment for more information.

## TODOS
- [x] Refactor messages to be localized. Use more generic message instead of ORCID special.
- [x] Test with ORCID (success!)
- [X] Test with Google (success!)
- [X] Test with Github (success!)

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

**Comment by pdurbin** [comment]
@poikilotherm as we're discussing at http://irclog.iq.harvard.edu/dataverse/2019-07-08#i_99911 I just left a comment at https://github.com/IQSS/dataverse/issues/5991#issuecomment-509177309 . I just took a quick look at the code and it's not immediately clear why the necessary refactoring is necessary. Can you please explain?

I'm going to leave this in code review and try to get some "talk after" time after standup.

**Comment by poikilotherm** [comment]
The refactoring is necessary due to upstream changes. ScribeJava has moved some classes around, deprecated (and dropped) some things usable with v3.1.0.
I also wanted to make the code more testable and easier to reuse, as there where some code dups (e. g. for the ORCiD part).
If you guys need a more elaborated answer, I am happy to open a call... Lot's of places things needed moving, too much to write down IMHO.

**Comment by poikilotherm** [comment]
I just removed the "WIP" from the title as requested by @pdurbin on IRC.
My refactoring is done for now, although this still lacks tests. I still need feedback on the code.

**Comment by djbrooke** [comment]
@poikilotherm - I assigned this to myself at standup and we'll put this on hold until I can review #5974 with the team.

**Comment by djbrooke** [comment]
@sekmiller @pdurbin @kcondon and I talked about this post-standup in regards to testing. Testing should be focused on whether the OAuth methods (ORCID, Github, Google) still work.

**Comment by poikilotherm** [comment]
I am happy to do the heavy lifting with this and share my test results. I hoped for some technical discussion first about design and usability of the new code.

**Comment by djbrooke** [comment]
Thanks @poikilotherm for the offer to help with the heavy lifting. @sekmiller is reviewing the code and will be able to provide any feedback. Is there anything specific about the approach that you'd like guidance on?

**Comment by djbrooke** [comment]
@sekmiller mentioned at standup that he'd like someone with more familiarity with the auth code to take a look at what's done so far. @scolapasta I'll assign to you so you can either be that person or work with that person to keep this moving. Thank you!

**Comment by poikilotherm** [comment]
Looks like you guys like my refactoring... I will test this with Github, Google and maybe ORCID, so please move this back to Community Dev column. Testing will not happen till next week.

**Comment by poikilotherm** [comment]
Thx @djbrooke

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java [hunk 1]
```diff
@@ -91,48 +95,80 @@ public String toString() {
...
-                                        parsed.username, 
-                                        OAuth2TokenData.from(accessToken),
-                                        parsed.displayInfo,
-                                        parsed.emails);
+        String body = response.getBody();
+        logger.log(Level.FINE, "In requestUserRecord. Body: {0}", body);
+        if ( responseCode == 200 && body != null ) {
+            return getUserRecord(body, accessToken, service);
         } else {
             throw new OAuth2Exception(responseCode, body, "Error getting the user info record.");
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java]: Is this error user facing? Does it show up in the web interface? If so it should be moved into Properties.bundle at some point (though maybe not in this pull request if we would like to defer this refactoring).

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java [hunk 2]
```diff
@@ -56,20 +62,78 @@
...
+                    String prettyUrl = response.encodeRedirectURL(destination);
+                    FacesContext.getCurrentInstance().getExternalContext().redirect(prettyUrl);
+                }
+            }
+        } catch (OAuth2Exception ex) {
+            error = ex;
+            logger.log(Level.INFO, "OAuth2Exception caught. HTTP return code: {0}. Message: {1}. Message body: {2}", new Object[]{error.getHttpReturnCode(), error.getLocalizedMessage(), error.getMessageBody()});
+            Logger.getLogger(OAuth2LoginBackingBean.class.getName()).log(Level.SEVERE, null, ex);
+        } catch (InterruptedException | ExecutionException ex) {
+            error = new OAuth2Exception(-1, "Please see server logs for more details", "Could not login due to threading exceptions.");
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java]: Same question as above. Does this English ever show up in the web interface for the end user?

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java [hunk 3]
```diff
@@ -79,60 +143,36 @@ public void exchangeCodeForToken() throws IOException {
...
-                tokenData.setUser(dvUser);
-                tokenData.setOauthProviderId(idp.getId());
-                oauth2Tokens.store(tokenData);
-                String destination = redirectPage.orElse("/");
-                HttpServletResponse response = (HttpServletResponse) FacesContext.getCurrentInstance().getExternalContext().getResponse();
-                String prettyUrl = response.encodeRedirectURL(destination);
-                FacesContext.getCurrentInstance().getExternalContext().redirect(prettyUrl);
+                return Optional.empty();
+            } catch (IOException e) {
+                error = new OAuth2Exception(-1, "", "Could not parse OAuth2 code due to IO error.");
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java]: More English.

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java [hunk 4]
```diff
@@ -79,60 +143,36 @@ public void exchangeCodeForToken() throws IOException {
...
+                logger.log(Level.WARNING, "IOException getting code parameter.", e.getLocalizedMessage());
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java]: Logging in English is fine. We can leave this alone.

**Comment by poikilotherm** [comment]
Notable changes for QA (WIP, more to come):
- Removed `scope` attribute from `OAuth2TokenData` in 81b367b7. See commit comment for more information.
- ...

**Comment by poikilotherm** [comment]
Alright, I tested this refactored code against GitHub, Google and ORCID sandbox. It all works flawless, affiliations are read from ORCID. Should be fine.

Just those text strings left over.

**Comment by pdurbin** [comment]
@poikilotherm thanks for keeping those strings in mind. When you're ready for this to go back into code review, please let us know!

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java (hunk 2 above)
Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java]: Most certainly not. Won't change...

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java (hunk 4 above)
Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java]: :+1:

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java (hunk 3 above)
Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java]: This exception is never thrown, but logged as localized message. Nothing TODO :smile:

File: src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java (hunk 1 above)
Comment (poikilotherm) [reply src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/AbstractOAuth2AuthenticationProvider.java]: Should be fixed in c729a27

**Comment by poikilotherm** [comment]
All set. Good night guys :sleeping:

**Comment by poikilotherm** [comment]
Thank you guys (@pdurbin, @scolapasta, @djbrooke, @kcondon et al) for merging.

As always, it has been a pleasure :smile:

**Pull request merged**